src/libraries/System.Private.CoreLib/src/System/Threading/Tasks/Sources/ManualResetValueTaskSourceCore.cs

```diff
@@ -209,49 +209,65 @@ private void SignalCompletion()
 
             if (_continuation != null || Interlocked.CompareExchange(ref _continuation, ManualResetValueTaskSourceCoreShared.s_sentinel, null) != null)
             {
-                if (_executionContext != null)
+                if (_executionContext != null) // ExecutionContext flow enabled
                 {
-                    ExecutionContext.RunInternal(
-                        _executionContext,
-                        (ref ManualResetValueTaskSourceCore<TResult> s) => s.InvokeContinuation(),
-                        ref this);
-                }
-                else
-                {
-                    InvokeContinuation();
-                }
-            }
-        }
+                    ExecutionContext? currentContext = ExecutionContext.Capture();
+                    // Restore the ExecutionContext before executing anything.
+                    ExecutionContext.Restore(_executionContext);
 
-        /// <summary>
-        /// Invokes the continuation with the appropriate captured context / scheduler.
-        /// This assumes that if <see cref="_executionContext"/> is not null we're already
-        /// running within that <see cref="ExecutionContext"/>.
-        /// </summary>
-        private void InvokeContinuation()
-        {
-            Debug.Assert(_continuation != null);
-
-            switch (_capturedContext)
-            {
-                case null:
-                    if (RunContinuationsAsynchronously)
+                    ExceptionDispatchInfo? edi = null;
+                    if (_capturedContext == null) // Default scheduler
                     {
-                        if (_executionContext != null)
+                        if (RunContinuationsAsynchronously)
                         {
                             ThreadPool.QueueUserWorkItem(_continuation, _continuationState, preferLocal: true);
                         }
                         else
                         {
-                            ThreadPool.UnsafeQueueUserWorkItem(_continuation, _continuationState, preferLocal: true);
+                            // Running inline may throw; capture the edi if it does.
+                            edi = InvokeContinuationInline();
                         }
                     }
+                    else // Custom scheduler
+                    {
+                        InvokeSchedulerContinuation();
+                    }
+
+                    // Restore the current context before propagating the throw; if there is one.
+                    ExecutionContext.Restore(currentContext);
+                    // Now rethrow the exception.
+                    edi?.Throw();
+                }
+                else if (_capturedContext == null) // Default scheduler, ExecutionContext flow suppressed
+                {
```

Comment:
This is the common case.  Maybe restructure things a bit so that the common cases come first? Just for reading purposes. e.g.
```C#
if (_executionContext is null)
{
    if (_capturedContext is null)
    {
         ... // most common case
         return;
    }

    ... // second most common case
    return;
}

if (_capturedContext is null)
{
    ... // rare
    return;
}

... // rarer still
return;
```

Reply:
Done